cpp-package/example/inference/sentiment_analysis_rnn.cpp

```diff
@@ -0,0 +1,397 @@
+    assert(false);
+  }
+
+  std::string line;
+  std::string word;
+  int index;
+  while (std::getline(fi, line)) {
+    std::istringstream stringline(line);
+    stringline >> word >> index;
+    wordToIndex[word] = index;
+  }
+  fi.close();
+}
+
+
+/*
+ * The function returns the index associated with the word in the dictionary.
+ * If the word is not present, the index representing "<unk>" is returned.
+ * If the "<unk>" is not present then 0 is returned.
+ */
+float Predictor::GetIndexForWord(const std::string& word) {
+  if (wordToIndex.find(word) == wordToIndex.end()) {
+    if (wordToIndex.find("<unk>") == wordToIndex.end())
+      return 0;
+    else
+      return static_cast<float>(wordToIndex["<unk>"]);
+  }
+  return static_cast<float>(wordToIndex[word]);
+}
+
+/*
+ * The function populates the input vector with indices from the dictionary that
+ * correspond to the words in the input string.
+ */
+int Predictor::ConverToIndexVector(const std::string& input, std::vector<float> *input_vector) {
+  std::istringstream input_string(input);
+  input_vector->clear();
+  const char delimiter = ' ';
+  std::string token;
+  size_t words = 0;
+  while (std::getline(input_string, token, delimiter) && (words <= input_vector->size())) {
+    input_vector->push_back(GetIndexForWord(token));
+    words++;
+  }
+  return words;
+}
+
+
+/*
+ * The function returns the index at which the given symbol name will appear
+ * in the output vector of NDArrays obtained after running the forward pass on the executor.
+ */
+int Predictor::GetIndexForOutputSymbolName(const std::string& output_symbol_name) {
+  int index = 0;
+  for (const std::string op : net.ListOutputs()) {
+    if (op == output_symbol_name) {
+      return index;
+    } else {
+      index++;
+    }
+  }
+  throw std::runtime_error("The output symbol name can not be found");
+}
+
+
+/*
+ * The following function runs the forward pass on the model.
+ * The executor is created in the constructor.
+ */
+float Predictor::PredictSentiment(const std::string& input_text) {
+  /*
+   * Initialize a vector of length equal to 'num_words' with index corresponding to <eos>.
+   * Convert the input string to a vector of indices that represent
+   * the words in the input string.
+   */
+  std::vector<float> index_vector(num_words, GetIndexForWord("<eos>"));
+  int num_words = ConverToIndexVector(input_text, &index_vector);
+
+  executor->arg_dict()["data0"].SyncCopyFromCPU(index_vector.data(), index_vector.size());
+  executor->arg_dict()["data1"] = num_words;
+
+  // Run the forward pass.
+  executor->Forward(false);
+
+  /*
+   * The output is available in executor->outputs. It is a vector of
+   * NDArray. We need to find the index in that vector that
+   * corresponds to the output symbol "sentimentnet0_hybridsequential0_dense0_fwd_output".
+   */
+  const std::string output_symbol_name = "sentimentnet0_hybridsequential0_dense0_fwd_output";
```

Comment:
This will cause problem for users trying to copy this code for their usecase or using a different model files. Can we think of simplifying this or making it more clear?

Reply:
The Forward() function returns a vector of NDArray each corresponding to output symbols.
Without knowing the name of symbol that is going to output the score, we can not find the exact index in that vector.
As long as users continue to use the same model, the symbol name won't change.
If users change the model, they would need to know the name of output symbol.

Reply:
I think your model only has one output, right?  [I verified by downloading model and confirming it only has one output]

So the ListOutputs() function only return a single element. You don't need to provide an explicit name, you can just use the one and single output element.

(It's true that if someone else copies this, and *their* model has multiple output elements, they would need to change the logic.)